contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,443 @@
+the driver recovers we should be able to resume gracefully.
+
+We will also create a playbook entry for how to turn off the CSI Driver
+migration gracefully, how to tell when the CSI Driver is broken or non-existent,
+and how to redeploy a CSI Driver in a cluster.
+
+
+## Upgrade/Downgrade, Migrate/Un-migrate
+### Kubelet Node Annotation
+When the Kubelet starts, it will check whether the feature gate is
+enabled and if so will annotate its node with `csi.attach.kubernetes.io/gce-pd`
+for example to communicate to the A/D Controller that it supports migration of
+the gce-pd to CSI. The A/D Controller will have to choose on a per-node basis
+whether to use the CSI or the in-tree plugin for attach based on 3 criterea:
+1. Feature gate
+2. Plugin Migratable (Implements MigratablePlugin interface)
+3. Node to Attach to has requisite Annotation
+
+Note: All 3 criterea must be satisfied for A/D controller to Attach/Detach with
+CSI instead of in-tree plugin. For example if a Kubelet has feature on and marks
+the annotation, but the A/D Controller does not have the feature gate flipped,
+we consider this user error and will throw some errors.
+
+This can cause a race between the A/D Controller and the Kubelet annotating, if
+a volume is attached before the Kubelet completes annotation the A/D controller
+could attach using in-tree plugin instead of CSI while the Kubelet is expecting
+a CSI Attach. The same issue exists on downgrade if the Annotation is not
+removed before a volume is attached. An additional consideration is that we
+cannot have the Kubelet downgraded to a version that does not have the
+Annotation removal code.
+
+### Node Drain Requirement
+We require node's to be drained whenever the Kubelet is Upgrade/Downgraded or
+Migrated/Unmigrated to ensure that the entire volume lifecycle is maintained
+inside one code branch (CSI or In-tree). This simplifies upgrade/downgrade
+significantly and reduces chance of error and races.
+
+### Upgrade/Downgrade Migrate/Unmigrate Scenarios
+For upgrade, starting from a non-migrated cluster you must turn on migration for
+A/D Controller first, then drain your node before turning on migration for the
+Kubelet. The workflow is as follows:
+1. A/D Controller and Kubelet are both not migrated
+2. A/D Controller restarted and migrated (flags flipped)
+3. A/D Controller continues to use in-tree code for this node b/c node
+   annotation doesn't exist
+4. Node drained, all volumes unmounted/detached with in-tree code
+5. Kubelet restarted and migrated (flags flipped)
+6. Kubelet annotates node to tell A/D controller this node has been migrated
+7. Both A/D Controller & Kubelet Migrated, node is in "fresh" state so all new
+   volumes lifecycle is CSI
+
+For downgrade, starting from a fully migrated cluster you must drain your node
+first, then turn off migration for your Kubelet, then turn off migration for the
+A/D Controller. The workflow is as follows:
+1. A/D Controller and Kubelet are both migrated
```

Comment:
The real sequence is probably something like this:

1. A/D Controller and Kubelet are both migrated
2. Kubelet drained and made unschedulable, all volumes unmounted/detached with CSI code
3. Kubelet restarted and un-migrated (flags flipped)
4. Kubelet removes node annotation to tell A/D Controller this node is not migrated. In case kubelet does not have annotation removal code, admin must remove the annotation manually.
5. Kubelet is made schedulable.
6. At this point all volumes going onto the node would be using in-tree code for both A/D Controller(b/c of annotation) and Kublet
6. Restart and un-migrate A/D Controller

Reply:
done